### PR TITLE
RGW: WIP D4N Filter

### PR DESCRIPTION
**Pull request by samarahu**
The D4N filter includes classes that inherit from the filter code with overriding methods that use D4N directory operations. It is turned on by default but requires the argument `-o rgw_filter=d4n` to be passed to the vstart call.  

Signed-off-by: Samarah Uriarte suriarte@redhat.com

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_directory.cc [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    client.sync_commit(std::chrono::milliseconds(1000));
+  }
+  catch(std::exception &e) {
+    exist = 0;
+  }
+
+  if (!exist) {
+    std::vector<std::pair<std::string, std::string>> list;
+    
+    if (host == "" || port == 0) {
```
Comment (samarahu) [inline src/rgw/rgw_directory.cc:79]: Does this check to ensure the endpoint is set make sense?

File: src/rgw/rgw_directory.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    RGWBlockDirectory() {}
+    RGWBlockDirectory(std::string blockHost, int blockPort):host(blockHost), port(blockPort) {}
+    
+    void init(CephContext *_cct) {
+      cct = _cct;
+      host = cct->_conf->rgw_directory_host;
+      port = cct->_conf->rgw_directory_port;
+    }
+	
+    void findClient(cpp_redis::client *client);
```
Comment (samarahu) [inline src/rgw/rgw_directory.h:44]: `client` is already a member variable, so it's redundant to pass it to this member function. As a result, should I remove this parameter? The only use I can see it having is if `findClient` will be called to find connected clients that do not belong to any `RGWBlockDirectory` instance.

File: src/rgw/rgw_sal_d4n.cc [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  std::unique_ptr<DeleteOp> d = next->get_delete_op();
+  return std::make_unique<D4NFilterDeleteOp>(std::move(d), this);
+}
+
+int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefixProvider* dpp)
+{
+  int getReturn = source->filter->get_block_dir()->getValue(source->filter->get_cache_block());
+
+  if (getReturn < 0) {
+    ldpp_dout(dpp, 20) << "D4N Filter: Directory get operation failed." << dendl;
```
Comment (samarahu) [inline src/rgw/rgw_sal_d4n.cc:81]: What logging level is best in this case?

Comment (mattbenjamin) [reply src/rgw/rgw_sal_d4n.cc:81]: if this even should "never happen" (it's a serious error) then level 0.  If it is relatively normal (we tried to get something but it doesn't happen to be cached), some level from 7 up to 20, lower if this is something basic that would be useful to a large number of debugging scenarios.

**Pull request closed without merging**